pkg/virt-api/rest/subresource.go

```diff
@@ -962,6 +940,61 @@ func generateVMVolumeRequestPatch(vm *v1.VirtualMachine, volumeRequest *v1.Virtu
 	return patch, nil
 }
 
+func getPatchVerb(requests []v1.VirtualMachineVolumeRequest) string {
+	verb := "add"
+	if len(requests) > 0 {
+		verb = "replace"
+	}
+	return verb
+}
+
+func addAddVolumeRequests(vm *v1.VirtualMachine, volumeRequest *v1.VirtualMachineVolumeRequest, vmCopy *v1.VirtualMachine) error {
+	name := volumeRequest.AddVolumeOptions.Name
+	for _, request := range vm.Status.VolumeRequests {
+		if err := validateAddVolumeRequest(request, name); err != nil {
+			return err
+		}
+	}
+	vmCopy.Status.VolumeRequests = append(vm.Status.VolumeRequests, *volumeRequest)
+	return nil
+}
+
+func validateAddVolumeRequest(request v1.VirtualMachineVolumeRequest, name string) error {
+	if shouldAddVolumeRequest(request, name) {
+		return fmt.Errorf("add volume request for volume [%s] already exists", name)
+	}
+	if shouldRemoveVolumeRequest(request, name) {
+		return fmt.Errorf("unable to add volume since a remove volume request for volume [%s] already exists and is still being processed", name)
+	}
+	return nil
+}
+
+func addRemoveVolumeRequests(vm *v1.VirtualMachine, volumeRequest *v1.VirtualMachineVolumeRequest, vmCopy *v1.VirtualMachine) error {
+	name := volumeRequest.RemoveVolumeOptions.Name
+	var volumeRequestsList []v1.VirtualMachineVolumeRequest
+	for _, request := range vm.Status.VolumeRequests {
+		if shouldAddVolumeRequest(request, name) {
+			// Filter matching AddVolume requests from the new list.
+			continue
+		}
+		if shouldRemoveVolumeRequest(request, name) {
+			return fmt.Errorf("a remove volume request for volume [%s] already exists and is still being processed", name)
+		}
+		volumeRequestsList = append(volumeRequestsList, request)
+	}
+	volumeRequestsList = append(volumeRequestsList, *volumeRequest)
+	vmCopy.Status.VolumeRequests = volumeRequestsList
+	return nil
+}
+
+func shouldRemoveVolumeRequest(request v1.VirtualMachineVolumeRequest, name string) bool {
```

Comment:
This function checks whether there already exists a request to remove the said volume, so we're not asking here whether or not we should remove a volume. IIUC, it's rather a question of `should we add a request that adds a volume?`. Therefore I would suggest naming this `removeVolumeRequestExists`